File: js/BaseModule.js
/**
 * Abstract Base Class for all AI Modules in JS.
 */
class BaseModule {
    constructor(name, config = {}) {
        if (this.constructor === BaseModule) {
            throw new Error("BaseModule is abstract and cannot be instantiated directly.");
        }
        this.name = name;
        this.config = config;
        this.enabled = config.enabled !== false;
    }

    /**
     * Process frame data and return results.
     */
    async process(videoElement) {
        throw new Error("Method 'process()' must be implemented.");
    }

    /**
     * Draw visual overlays on the canvas.
     */
    draw(ctx, results, canvas) {
        throw new Error("Method 'draw()' must be implemented.");
    }
}

File: core/base_module.py
from abc import ABC, abstractmethod
import numpy as np

class BaseModule(ABC):
    """
    Abstract Base Class for all AI Modules.
    Ensures a consistent interface across the plugin-based system.
    """

    def __init__(self, name: str, config: dict = None):
        self.name = name
        self.config = config or {}
        self.enabled = self.config.get('enabled', True)

    @abstractmethod
    def process(self, frame: np.ndarray) -> dict:
        """
        Process the input frame and return results as a dictionary.
        """
        pass

    @abstractmethod
    def draw(self, frame: np.ndarray, results: dict) -> np.ndarray:
        """
        Draw visual overlays on the frame based on process results.
        """
        pass
